### ISSUE TITLE
Dapr sidecar cannot start with nextjs application

### ISSUE THREAD

**Issue by Garwakizashi** [microsoft/azure-container-apps#1182]
### This issue is a: (mark with an x)
- [ x] bug report 
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)

### Issue description
Not sure if this can be classed as a bug, as it is very likely I'm missing on my end to get this working; but I'd like to get some guidance with this issue I'm having.

I'm trying to enable dapr on an ACA that runs a nextjs application. 

The application itself is running on port `3000`, and that is also what I've set the app-port of dapr to be as well. However, the the container just keeps crashing, looking at the logs of daprd, it just shows the debug message several times:
```
level=debug msg="api error: code = Internal desc = dapr is not ready"
```
before eventually:
```
level=info msg="Received signal 'terminated'; beginning shutdown" 
```

I did notice that it said it is waiting on the port, but the application never gets discovered. Could this be why it is failing?
Is there a requirement on my side to set up in nextjs to allow dapr to discover it? The ACA ingress port is also set to `3000`, and is working as expected. I have several ACA's that are working fine with dapr enabled, so not sure why this particular app would fail.
Unsure if this helps, but I've been able to run dapr locally and the application gets discovered fine.

### Steps to reproduce
1. Deploy a nextjs application on ACA
2. Enable Dapr 

**Expected behavior**
Daprd container to work.
**Actual behavior**
Daprd container keeps restarting due to the described error above.

### Additional context
Sample logs from log stream (removed instance name):
``` 
2024-05-29T22:14:14.61413  Connecting to the container 'daprd'...
2024-05-29T22:14:14.63342  Successfully Connected to container: 'daprd'
2024-05-29T22:11:33.722306026Z time="2024-05-29T22:11:33.722256786Z" level=info msg="Enabled gRPC tracing middleware" app_id=test scope=dapr.runtime.grpc.internal type=log ver=1.12.5
2024-05-29T22:11:33.722310735Z time="2024-05-29T22:11:33.722267235Z" level=info msg="Enabled gRPC metrics middleware" app_id=test scope=dapr.runtime.grpc.internal type=log ver=1.12.5
2024-05-29T22:11:33.722351340Z time="2024-05-29T22:11:33.722302922Z" level=info msg="Internal gRPC server is running on port 50002" app_id=test scope=dapr.runtime type=log ver=1.12.5
2024-05-29T22:11:33.722363443Z time="2024-05-29T22:11:33.722328089Z" level=info msg="application protocol: http. waiting on port 3000.  This will block until the app is listening on that port." app_id=test scope=dapr.runtime type=log ver=1.12.5
2024-05-29T22:11:36.882162928Z time="2024-05-29T22:11:36.881484614Z" level=debug msg="api error: code = Internal desc = dapr is not ready" app_id=test scope=dapr.runtime.http type=log ver=1.12.5
2024-05-29T22:11:36.882186211Z time="2024-05-29T22:11:36.881724551Z" level=debug msg="api error: code = Internal desc = dapr is not ready" app_id=test  scope=dapr.runtime.http type=log ver=1.12.5
2024-05-29T22:11:42.881653629Z time="2024-05-29T22:11:42.881480949Z" level=debug msg="api error: code = Internal desc = dapr is not ready" app_id=test  scope=dapr.runtime.http type=log ver=1.12.5
2024-05-29T22:11:42.881991531Z time="2024-05-29T22:11:42.881895281Z" level=debug msg="api error: code = Internal desc = dapr is not ready" app_id=test  scope=dapr.runtime.http type=log ver=1.12.5
2024-05-29T22:11:48.882059272Z time="2024-05-29T22:11:48.881965535Z" level=debug msg="api error: code = Internal desc = dapr is not ready" app_id=test scope=dapr.runtime.http type=log ver=1.12.5
2024-05-29T22:11:48.882360073Z time="2024-05-29T22:11:48.88213447Z" level=debug msg="api error: code = Internal desc = dapr is not ready" app_id=test scope=dapr.runtime.http type=log ver=1.12.5
2024-05-29T22:11:48.883884885Z time="2024-05-29T22:11:48.883683019Z" level=debug msg="api error: code = Internal desc = dapr is not ready" app_id=test  scope=dapr.runtime.http type=log ver=1.12.5
2024-05-29T22:11:48.891698249Z time="2024-05-29T22:11:48.891019754Z" level=info msg="Received signal 'terminated'; beginning shutdown" app_id=test  scope=dapr.signals type=log ver=1.12.5
2024-05-29T22:11:48.891709700Z time="2024-05-29T22:11:48.891051904Z" level=debug msg="stopping workload cert expiry watcher" app_id=test scope=dapr.runtime.security type=log ver=1.12.5
2024-05-29T22:11:48.891714890Z time="2024-05-29T22:11:48.891163221Z" level=debug msg="Graceful shutdown timeout: 5s" app_id=test scope=dapr.concurrency.closer type=log ver=1.12.5
2024-05-29T22:11:48.891719438Z time="2024-05-29T22:11:48.891202174Z" level=info msg="Dapr is shutting down" app_id=test  scope=dapr.runtime type=log ver=1.12.5
2024-05-29T22:11:48.891724207Z time="2024-05-29T22:11:48.891228814Z" level=info msg="Dapr runtime stopped" app_id=test scope=dapr.runtime type=log ver=1.12.5
2024-05-29T22:11:48.891730378Z time="2024-05-29T22:11:48.89124825Z" level=info msg="Shutting down workflow engine" app_id=test scope=dapr.runtime type=log ver=1.12.5
2024-05-29T22:11:48.891734857Z time="2024-05-29T22:11:48.891355159Z" level=info msg="Closing HTTP server :3501…" app_id=test scope=dapr.runtime.http type=log ver=1.12.5
2024-05-29T22:11:48.891739065Z time="2024-05-29T22:11:48.891432924Z" level=info msg="Closing HTTP server [::1]:3500…" app_id=test  scope=dapr.runtime.http type=log ver=1.12.5
2024-05-29T22:11:48.891743823Z time="2024-05-29T22:11:48.891477347Z" level=info msg="Closing HTTP server 127.0.0.1:3500…" app_id=test scope=dapr.runtime.http type=log ver=1.12.5
```

**Comment by BlackRider97** [comment]
@Garwakizashi There is no known issue with Dapr integration with ACA. We have variety of Apps running in ACA and could not see any reason why it won't work for you.  Can you please raise a support ticket with more details about your App Env and other details to debug this better? Please link this GitHub issue to be aware of the context.

**Comment by Garwakizashi** [comment]
> @Garwakizashi There is no known issue with Dapr integration with ACA. We have variety of Apps running in ACA and could not see any reason why it won't work for you. Can you please raise a support ticket with more details about your App Env and other details to debug this better? Please link this GitHub issue to be aware of the context.

Thanks for the reply. I wasn't able to figure why it wouldn't work, but I managed to get around it by running the nextjs app behind nginx. Once it was behind nginx, the dapr sidecar started successfully.

**Comment by bfigueir** [comment]
+1

Also facing this issue with ACA. Locally I can run de Nextjs app with Dapr without issues. When deploying to ACA, Dapr gives the same error has stated by @Garwakizashi

**Comment by BlackRider97** [comment]
@bfigueir Are you configuring `dapr.appPort` correctly inside container app template? Please note that it must be same as `ingress.targetPort`

<img width="662" alt="Image" src="https://github.com/user-attachments/assets/8d97a2e8-c3d9-40da-aaa1-817e6ddfaef5" />
https://learn.microsoft.com/en-us/azure/container-apps/enable-dapr?tabs=bicep1

More details [here](https://learn.microsoft.com/en-us/rest/api/resource-manager/containerapps/container-apps/create-or-update?view=rest-resource-manager-containerapps-2024-03-01&tabs=HTTP)

**Comment by bfigueir** [comment]
Hi @BlackRider97 yes, we correctly mapping the ports. 
We have a container environment with several container apps, in a mix of nextjs apps and fastify services. We are only hitting this issue with the nextjs ones. We had to do the same that @Garwakizashi and run the nextjs behind a NGINX acting as a reverse proxy.
Thanks

**Comment by Garwakizashi** [comment]
Just to continue this thread as it is dependant on my reply; yes, I've mapped the correct port on the container app. I'm assuming that @bfigueir  had the same logs as I did when we tried to deploy the nextjs app without NGINX.

**Comment by jleight** [comment]
I ran into this same issue today. After seeing people recommending running their app behind nginx, I decided to try the same thing with HAProxy. Turns out, HAProxy ended up having the same problem that we're seeing with Dapr.

My HAProxy config was pointing at `127.0.0.1:3001` (I had to change the Next.js app to port 3001 so HAProxy could be on port 3000), and it was unable to talk to the app. I reconfigured HAProxy to use `"$HOSTNAME":3001` as the backend and that fixed the issue.

Since I still wasn't happy about using a reverse proxy in front of the Next.js app, I looked into the `server.js` that Next.js was building and found:

```
const hostname = process.env.HOSTNAME || '0.0.0.0'
```

I used `sed` in my dockerfile to get rid of the `process.env.HOSTNAME ||`:

```
RUN sed -i 's/process\.env\.HOSTNAME || //' /path/to/.next/standalone/server.js
```

... and now it works fine with Dapr and without an extra reverse proxy.

**Comment by brunoshine** [comment]
I can confirm that @jleight solution works in ACA. thanks @jleight nice work 🥇

**Comment by xaullepaul** [comment]
I can confirm that @jleight solution works, tested in app service with cloudflared sidecar. And also I want to add, instead of removing the process.env.HOSTNAME, you can explicitly add environment variable HOSTNAME=0.0.0.0 that it'll set the hostname to be attach on app service 0.0.0.0:port and appcontainer localhost:port.  It will not randomly create container name anymore.

Then you can call the service directly localhost:port same as other sidecar apps.

**Comment by BlackRider97** [comment]
I’m not very familiar with the Next.js stack, so if you can share a public container image, it would help me reproduce the issue. My initial suspicion is that this might be related to a specific Next.js configuration.

@xaullepaul  @brunoshine @Garwakizashi

**Comment by BlackRider97** [comment]
The community has identified the root cause and solution for this issue — thanks to @jleight, @xaullepaul, and @brunoshine for the investigation! 🎉

**Root Cause:** Next.js standalone mode sets `hostname = process.env.HOSTNAME || '0.0.0.0'`. In ACA, the `HOSTNAME` env var is set to the container hostname (not `0.0.0.0`), so the app only listens on that specific hostname. Dapr connects to the app via `127.0.0.1`, causing a connection mismatch.

**Solution:** Set the environment variable `HOSTNAME=0.0.0.0` on your container app:
```bash
az containerapp update --name <app> --resource-group <rg> --set-env-vars "HOSTNAME=0.0.0.0"
```
This ensures Next.js listens on all interfaces, allowing Dapr's localhost connection to succeed.

Closing with this documented solution. If the `HOSTNAME` env var approach doesn't work for your setup, please open a new issue with details.

---
*🤖 This comment was generated with AI assistance and may contain inaccuracies. If this doesn't address your concern or if the issue persists, please reopen this issue or create a new one.*
